File: rpidash/services/current_utilization.py
# THIRD PARTY
from flask.typing import ResponseReturnValue
from flask.views import View

# FIRST PARTY
from rpidash.utils import (
    get_cpu_temperature,
    get_cpu_utilization,
    get_memory_utilization,
    get_storage_utilization,
)


class CurrentUtilization(View):
    """Current system utilization view."""

    def dispatch_request(self) -> ResponseReturnValue:
        """Render current system utilization API view."""
        memory_percentage, memory_used, memory_total = get_memory_utilization()
        storage_percentage, storage_used, storage_total = get_storage_utilization()
        response = {
            "cpu_temperature": get_cpu_temperature(),
            "cpu_percentage": get_cpu_utilization(),
            "memory_percentage": memory_percentage,
            "memory_used": memory_used,
            "memory_total": memory_total,
            "storage_percentage": storage_percentage,
            "storage_used": storage_used,
            "storage_total": storage_total,
        }
        return response

File: rpidash/static/js/current-utilization.js
const fetchCurrentUtilization = async () => {
  try {
    const response = await fetch("/services/current_utilization");
    if (!response.ok) {
      throw new Error('Network response was not ok');
    }
    const data = await response.json();
    displayCurrentUtilization(data);
  } catch (error) {
    console.error("There was a problem fetching the data:", error);
  }
};

const displayCurrentUtilization = ({
  cpu_percentage,
  cpu_temperature,
  memory_used,
  memory_percentage,
  memory_total,
  storage_used,
  storage_percentage,
  storage_total
}) => {
  const cpuUtilizationElement = document.getElementById("current-cpu-utilization");
  const cpuTemperatureElement = document.getElementById("current-cpu-temperature");
  const memoryUtilizationElement = document.getElementById("current-memory-utilization");
  const storageUtilizationElement = document.getElementById("current-storage-utilization");

  cpuUtilizationElement.textContent = `${cpu_percentage}%`;
  cpuTemperatureElement.textContent = `${cpu_temperature} °C`;
  memoryUtilizationElement.textContent = `${memory_used} MB (${memory_percentage}%) of ${memory_total} MB`;
  storageUtilizationElement.textContent = `${storage_used} GB (${storage_percentage}%) of ${storage_total} GB`;

  const cpuUtilizationProgress = document.getElementById("cpu-utilization-progress");
  const cpuTemperatureProgress = document.getElementById("cpu-temperature-progress");
  const memoryUtilizationProgress = document.getElementById("memory-utilization-progress");
  const storageUtilizationProgress = document.getElementById("storage-utilization-progress");

  cpuUtilizationProgress.style.width = `${100 - cpu_percentage}%`;
  cpuTemperatureProgress.style.width = `${100 - cpu_temperature}%`;
  memoryUtilizationProgress.style.width = `${100 - memory_percentage}%`;
  storageUtilizationProgress.style.width = `${100 - storage_percentage}%`;
};

fetchCurrentUtilization();
setInterval(fetchCurrentUtilization, 10000);